src/main/java/com/microsoft/azure/vmagent/AzureVMManagementServiceDelegate.java

```diff
@@ -385,14 +458,53 @@ public static void setVirtualMachineDetails(
         VirtualMachine vm = azureClient.virtualMachines().getByGroup(template.getResourceGroupName(), azureAgent.getNodeName());
 
         // Getting the first virtual IP
-        azureAgent.setPublicDNSName(vm.getPrimaryPublicIpAddress().fqdn());
+        final PublicIpAddress publicIP = vm.getPrimaryPublicIpAddress();
+        String publicIPStr = "";
+        String privateIP = vm.getPrimaryNetworkInterface().primaryPrivateIp();
+        String fqdn = "";
+        if (publicIP == null) {
+            fqdn = privateIP;
+            LOGGER.log(Level.INFO, "The Azure agent doesn't have a public IP. Will use the private IP");
+        } else {
+            fqdn = publicIP.fqdn();
+            publicIPStr = publicIP.ipAddress();
+        }
+        azureAgent.setPublicDNSName(fqdn);
         azureAgent.setSshPort(Constants.DEFAULT_SSH_PORT);
+        azureAgent.setPublicIP(publicIPStr);
+        azureAgent.setPrivateIP(privateIP);
 
         LOGGER.log(Level.INFO, "Azure agent details:\nnodeName{0}\nadminUserName={1}\nshutdownOnIdle={2}\nretentionTimeInMin={3}\nlabels={4}",
                 new Object[]{azureAgent.getNodeName(), azureAgent.getVMCredentialsId(), azureAgent.isShutdownOnIdle(),
                     azureAgent.getRetentionTimeInMin(), azureAgent.getLabelString()});
     }
 
+    public static void attachPublicIP(final AzureVMAgent azureAgent, final AzureVMAgentTemplate template) throws Exception {
+        final Azure azureClient = TokenCache.getInstance(template.getAzureCloud().getServicePrincipal()).getAzureClient();
+        final VirtualMachine vm = azureClient.virtualMachines().getByGroup(template.getResourceGroupName(), azureAgent.getNodeName());
+        LOGGER.log(Level.INFO, "Trying to attach a public IP to the agent {0}", azureAgent.getNodeName());
+        if (vm != null) {
+            //check if the VM already has a public IP
+            if (vm.getPrimaryPublicIpAddress() == null) {
+                vm.getPrimaryNetworkInterface()
+                        .update()
+                        .withNewPrimaryPublicIpAddress(
+                                azureClient.publicIpAddresses()
+                                        .define(azureAgent.getNodeName() + "IPName")
+                                        .withRegion(template.getLocation())
+                                        .withExistingResourceGroup(template.getResourceGroupName())
+                                        .withLeafDomainLabel(azureAgent.getNodeName())
+                        ).apply();
```

Comment:
apply() is synchronous, it doesn't return until the PIP is actually created. 
setVirtualMachineDetails asks Azure again so it should always get the PIP (I've tested empirically and it does - if it wont' then it's an SDK bug)